lua/orgmode/ui/virtual-indent.lua

```diff
@@ -0,0 +1,101 @@
+local Headline = require('orgmode.treesitter.headline')
+
+---@alias VirtualIndentHandler fun(buffer:number, start_line: number, end_line: number) function that sets the indentation
+
+---@class VirtualIndent
+---@field private _ns_id number extmarks namespace id
+---@field private _handler VirtualIndentHandler? function that sets the indentation
+local VirtualIndent = {}
+
+---@class VirtualIndentData
+---@field handler VirtualIndentHandler? function that sets the indentation
+
+---@param data VirtualIndentData
+function VirtualIndent:new(data)
+  data = data or {}
+
+  vim.validate({
+    handler = { data.handler, 'function', true },
+  })
+
+  local opts = {}
+  opts._ns_id = vim.api.nvim_create_namespace('orgmode.ui.indent')
+  opts._handler = data.handler
+
+  setmetatable(opts, self)
+  self.__index = self
+  return opts
+end
+
+function VirtualIndent:_delete_old_extmarks(buffer, start_line, end_line)
+  local old_extmarks = vim.api.nvim_buf_get_extmarks(
+    buffer,
+    self._ns_id,
+    { start_line, 0 },
+    { end_line, 0 },
+    { type = 'virt_text' }
+  )
+  for _, ext in ipairs(old_extmarks) do
+    vim.api.nvim_buf_del_extmark(buffer, self._ns_id, ext[1])
+  end
+end
+
+function VirtualIndent:_get_indent_size(line)
+  local headline = Headline.from_cursor({ line + 1, 1 })
+
+  if headline then
+    local level = headline:level()
+    local headline_line, _, _ = headline.headline:start()
+
+    if headline_line == line then
+      return level - 1
+    else
+      return level * 2
+    end
+  end
+
+  return 0
+end
+
+---@param buffer number buffer id
+---@param start_line number start line number to set the indentation, 0-based inclusive
+---@param end_line number end line number to set the indentation, 0-based inclusive
+function VirtualIndent:set_indent(buffer, start_line, end_line)
+  if self._handler then
+    return self._handler(buffer, start_line, end_line)
+  end
+
+  self:_delete_old_extmarks(buffer, start_line, end_line)
+  for line = start_line, end_line do
+    local indent = self:_get_indent_size(line)
+
+    if indent > 0 then
+      vim.api.nvim_buf_set_extmark(buffer, self._ns_id, line, 0, {
+        virt_text = { { string.rep(' ', indent), 'OrgIndent' } },
+        virt_text_pos = 'inline',
+        right_gravity = false,
+      })
+    end
+  end
+end
+
+---@param buffer number buffer id
+function VirtualIndent:attach(buffer)
+  vim.validate({
+    buffer = { buffer, 'number', true },
+  })
+
+  buffer = buffer or 0
+
+  self:set_indent(buffer, 0, vim.api.nvim_buf_line_count(buffer) - 1)
+
+  vim.api.nvim_buf_attach(buffer, true, {
```

Comment:
Why do we need `vim.schedule` in the callback?

Also, is there any advantage of using `nvim_buf_attach` over `nvim_set_decoration_provider` ?
We use decoration provider for markup, so we could maybe use the same for this while using the `ephermal` ?

Reply:
We need `vim.schedule`, because otherwise, when undoing, the virtual indent handler is called too early and processed incorrectly.

As far as I know, ephemeral virtual text has not been implemented yet.

Reply:
I gave it a quick test, and it seems to work. Switching to it should reduce some code that does the cleanup. Lets try switching to decoration provider and see if it works out.

Reply:
Is it true? I get `E5108: Error executing lua [string ":lua"]:1: not yet implemented` error.

Here is the code I run:
```lua
vim.api.nvim_buf_set_extmark(0, vim.api.nvim_create_namespace("test"), 0, 1, {virt_text_pos = "inline", virt_text = {{"    "}}, ephemeral = true})
```

My neovim version is `NVIM v0.10.0-dev-410+gc48f94d1f`

Reply:
Okay, I get the point. I'll try to redo it